--- magick/composite-private.h.orig	2015-06-02 03:46:03.914115735 +0200
+++ magick/composite-private.h	2015-06-02 03:46:36.947582226 +0200
@@ -58,6 +58,12 @@
   /*
     Compose pixel p over pixel q with the given opacities.
   */
+  if (fabs((double) (alpha-TransparentOpacity)) < MagickEpsilon)
+  {
+    *composite=(*q);
+    return;
+  }
+
   Sa=1.0-QuantumScale*alpha;
   Da=1.0-QuantumScale*beta;
 #if !defined(MAGICKCORE_HDRI_SUPPORT)
