Index: /ImageMagick/trunk/Magick++/lib/Magick++/Functions.h
===================================================================
--- Magick++/lib/Magick++/Functions.h	(revision 17507)
+++ Magick++/lib/Magick++/Functions.h	(revision 18632)
@@ -29,4 +29,7 @@
   MagickPPExport void InitializeMagick(const char *path_);
 
+  // Seed a new sequence of pseudo-random numbers
+  MagickPPExport void SetRandomSeed(const unsigned long seed);
+
   // C library initialization routine
   MagickPPExport void TerminateMagick();
Index: /ImageMagick/trunk/Magick++/lib/Functions.cpp
===================================================================
--- Magick++/lib/Functions.cpp	(revision 17621)
+++ Magick++/lib/Functions.cpp	(revision 18632)
@@ -59,4 +59,9 @@
 }
 
+MagickPPExport void Magick::SetRandomSeed(const unsigned long seed)
+{
+  MagickCore::SetRandomSecretKey(seed);
+}
+
 MagickPPExport void Magick::TerminateMagick(void)
 {
